Rich-Text-Editor-(RTE)-Overview.md

```diff
@@ -1,117 +1,137 @@
+* The function passed to `ck.on('change', ...)` is executed whenever the content of CKEditor changes. It handles user updates to the rich text.
+
+### Adding a new plugin
+
+To add a new plugin, follow these steps:
+
+1. Add the plugin to the `manifest.json` to download it. `SharedSpace` can be looked as an example. Alternatively, define your own custom plugin like we did with `pre`.
+2. Add a call to `CKEditor.plugins.addExternal` to the component's `ngAfterViewInit()` method.
+3. Add this plugin to `extraPlugins` in `ckConfig`.
+
+#### Adding rich text components to CKEditor
+
+Rich text components are added to CKEditor in [`ck-editor-4-widgets.initializer.ts`](https://github.com/oppia/oppia/blob/develop/core/templates/components/ck-editor-helpers/ck-editor-4-widgets.initializer.ts). The components are dynamically added to CKEditor as [widgets](https://docs.ckeditor.com/ckeditor4/latest/guide/widget_sdk_intro.html). We use the `getRichTextComponents()` function to obtain the information of each rich text component and construct there respective widgets. We also have a function `isInlineComponent()` to check whether a rich text component is inline or block component.
+
+`componentTemplate` defines a template to wrap the rich text components. Inline components are wrapped in a span, and block components are wrapped in a div.
+
+The plugins are added to CKEditor in the line `CKEDITOR.plugins.add(ckName, {`. The `init` function passed to `add()` is executed when the plugin is initialised and it adds a widget for each component using the line `editor.widgets.add(ckName, {`. `editor.widgits.add` is passed an object with the following attributes:
+
+* `button`: The button label for the widget.
+* `inline`: Whether the widget is an inline component or block component. The link and the math component are inline and the rest are declared block components.
+* `template`: The wrapper template for the widget.
+* `edit`: A function that will be executed when a widget is being edited. In this function the default action is canceled since we have used our own edit modal. `RteHelperService._openCustomizationModal()`is a helper function for opening the modal which is used to insert new components or editing existing ones. It uses the customizationArgSpecs (obtained from the component definition) to know what the editable fields are for each component, which allows it to render the modal properly for any component.
+* `downcast()`: A function that downcasts the widget instance by clearing the Angular rendering content and returning the rich text component without any wrapper. Downcasting a widget means removing any property which is not needed in the output's source.
```

Comment:
Perhaps explain when/whether we need to use these functions.

Reply:
These functions won't have to be used anywhere else in the codebase. This section isn't explaining how to do something; it's explaining how RTEs have already been implemented. Quoting from the wiki:

> This section is a code overview of how the RTE is actually implemented. This is mostly useful if you plan to modify the RTE when fixing a bug or adding a new feature.

I've added to the quoted paragraph to explain that this section is not needed for adding or using RTEs

Reply:
Sorry, I'm a bit confused -- that paragraph still looks the same as what you've quoted here. Could you please clarify what additions etc. were made? I re-read a couple of times but I might be missing them.

Reply:
Ahh sorry. I decided later to put the explanation under the "Implementation" heading instead. It's on line 130:

> This section describes how Oppia's rich text editor and rich text component system are implemented. This information may be helpful if you need to modify how Oppia uses rich text components, but you don't need this section if you just want to use RTEs or add new rich text components.